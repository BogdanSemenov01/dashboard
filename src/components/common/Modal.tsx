import React, { useContext, useEffect, useState } from "react"
import styled, { css, keyframes } from "styled-components"
import { ModalContext } from "../../context/ModalContext/ModalContext"

const showModalBackDrop = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`
const hideModalBackDrop = keyframes`
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
`

const showModal = keyframes`
  from {
    transform: scale(0)
  }

  to {
    transform: scale(1)
  }
`

const hideModal = keyframes`
  from {
    transform: scale(1)
  }

  to {
    transform: scale(0)
  }
`

const StyledModalBackDrop = styled("div")<{ closing: boolean }>`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.1);
  z-index: 100;

  ${(props) =>
    props.closing ? backdropAnimationOnClose : backdropAnimationOnOpen};
`

const StyledModal = styled("div")<{ closing: boolean }>`
  margin-top: -200px;
  background-color: white;
  width: 800px;
  height: 600px;
  border-radius: 10px;
  box-shadow: 0 0 25px rgba(0, 0, 0, 0.3);

  ${(props) => (props.closing ? modalAnimationOnClose : modalAnimationOnOpen)};
`

const modalAnimationOnOpen = css`
  animation: ${showModal} 0.3s ease-out;
`
const modalAnimationOnClose = css`
  animation: ${hideModal} 0.3s ease-out;
`
const backdropAnimationOnOpen = css`
  animation: ${showModalBackDrop} 0.3s ease-out;
`
const backdropAnimationOnClose = css`
  animation: ${hideModalBackDrop} 0.3s ease-out;
`

const StyledModalHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 20px;
  border-bottom: 1px solid rgb(219, 219, 219);
  & > h3 {
    margin: 0;
    padding: 0;
    font-size: 20px;
  }
`

const StyledModalBody = styled.div`
  padding: 10px 20px;
`

type ModalProps = {
  title: string
  children: React.ReactElement | undefined
  
}

const Modal = (props: ModalProps) => {
  const { title, children } = props

  const { closeModal } = useContext(ModalContext)

  const [closing, setClosing] = useState(false)

  const onClickClose = () => {
    setClosing(true)
    const closeTimeout = setTimeout(() => {
      clearTimeout(closeTimeout)
      closeModal()
    }, 290)
  }

  return (
    <StyledModalBackDrop onClick={onClickClose} closing={closing}>
      <StyledModal
        onClick={(e) => e.stopPropagation()}
        closing={closing}
        data-testid='modal'
      >
        <StyledModalHeader>
          <h3>{title}</h3>
          <button onClick={onClickClose}>x</button>
        </StyledModalHeader>
        <StyledModalBody>{children}</StyledModalBody>
      </StyledModal>
    </StyledModalBackDrop>
  )
}

export default Modal
