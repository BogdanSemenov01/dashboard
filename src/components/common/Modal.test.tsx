import { render, screen, waitFor } from "@testing-library/react";
import Modal from "./Modal";
import userEvent from "@testing-library/user-event";

const Element = () => {
  return <div>Test</div>  
}

describe('Modal window', () => {
  it('should render modal window component', () => {
    const component = render(<Modal title='Modal'><Element /></Modal>)
    expect(component).toMatchSnapshot()
  })

  it ('should call close function on close',  () => {
    render(<Modal title='Modal'><Element /></Modal>)
    const button = screen.getByRole('button')
    userEvent.click(button)


    expect(screen.queryByTestId('modal')).toBeNull()
  })

})