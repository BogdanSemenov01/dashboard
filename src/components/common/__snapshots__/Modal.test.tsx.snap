// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal window should render modal window component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe jbUiCN"
      >
        <div
          class="sc-gswNZR lhyzCb"
        >
          <div
            class="sc-dkrFOg jxUGFB"
          >
            <h3>
              Modal
            </h3>
            <button>
              x
            </button>
          </div>
          <div
            class="sc-hLBbgP iRXSkq"
          >
            <div>
              Test
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe jbUiCN"
    >
      <div
        class="sc-gswNZR lhyzCb"
      >
        <div
          class="sc-dkrFOg jxUGFB"
        >
          <h3>
            Modal
          </h3>
          <button>
            x
          </button>
        </div>
        <div
          class="sc-hLBbgP iRXSkq"
        >
          <div>
            Test
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
